File: LimSeoyeon/03_Recursion/BOJ_5904_Moo게임.java
import java.util.Scanner;

public class Main {
    static int idx = 0;
    static int N;
    static boolean flag = false;

    public static void recursion(int n) {
        if (n == 0) {
            if (idx == N) {
                System.out.println('m');
                flag = true;
                return;
            } else if (idx + 1 == N || idx + 2 == N) {
                System.out.println('o');
                flag = true;
                return;
            } else {
                idx += 3;
            }
            return;
        }
        recursion(n - 1);
        if (flag) return;

        if (idx == N) {
            System.out.println('m');
            flag = true;
            return;
        } else if (idx + 1 <= N && idx + n + 3 > N) {
            System.out.println('o');
            flag = true;
            return;
        } else {
            idx += n + 3;
        }
        recursion(n - 1);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        N = scanner.nextInt();
        N = N - 1;
        int cnt = 3;
        int i = 0;
        while (cnt <= N) {
            cnt = cnt * 2 + (i + 3);
            i++;
        }
        recursion(i);
    }
}

File: LimSeoyeon/03_Recursion/BOJ_5904_Moo게임.cpp
#include <iostream>
using namespace std;

int idx = 0;
char res;
bool flag = 0;
int N;

void recursion(int n) {
	if (flag) return;
	if (n == 0) {
		if (idx == N) {
			res = 'm';
			flag = 1;
		}
		else if (idx + 1 == N || idx + 2 == N) {
			res = 'o';
			flag = 1;
		}
		else {
			idx += 3;
		}
		return;
	}
	recursion(n - 1);
	if (idx == N) {
		res = 'm';
		flag = 1;
	}
	else if (idx+1 <= N && idx +n+3 > N) {
		res = 'o';
		flag = 1;
	}
	else {
		idx += n + 3;
	}
	recursion(n - 1);

}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> N;
	N = N - 1;
	int cnt = 3;
	int i = 0;
	while (cnt<=N) {
		cnt = cnt * 2 + (i+3);
		i++;
	}
	recursion(i);
	
	cout << res;
}
